# Question 1: Exercise Practice

## 01 Importing Libraties

In [1]:
import pandas as pd
import numpy as np
import os

## 02 Turn your project folder path into a string 

In [2]:
path = r'C:\Users\nanu_\Documents\Career Foundry\Projects\2 Data Immersion\Python fundamental for Data Analysis\01-2024 Instacart Basket analysis'

In [3]:
path

'C:\\Users\\nanu_\\Documents\\Career Foundry\\Projects\\2 Data Immersion\\Python fundamental for Data Analysis\\01-2024 Instacart Basket analysis'

## 03 Importing  orders data

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

## 04 Importing product Dataset

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## 05 Data wrangling

### Dropping eval_set column from orders.csv

In [6]:
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [7]:
df_ords_2

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

### counting the missing value in days_since_prior_order

In [8]:
df_ords_2['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

## 06 Renaming Columns

### Renaming order_dow column from orders table

In [9]:
df_ords_2.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
df_ords_2.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## 07 Changing a Variable's Data type

### Changing data type of  order_id in orders table

In [11]:
df_ords_2.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [12]:
df_ords_2['order_id'] = df_ords['order_id'].astype('str')

In [13]:
df_ords_2['order_id'].dtype

dtype('O')

## 08 Transposing Data

### Importing data set department.csv

In [14]:
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

### Transposing department table

In [16]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
df_dep_t = df_dep.T

In [18]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

### Adding index  column in Departments

In [19]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

### Creating a new header in Dataframe

In [20]:
new_header = df_dep_t.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

In [22]:
df_dep_t_new = df_dep_t[1:]

In [23]:
df_dep_t_new 

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [24]:
df_dep_t_new.columns = new_header

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## 09 Data dictionaries

In [26]:
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [28]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
print(data_dict.get('19'))

{'department': 'snacks'}


## 10 Subsetting

### creating subset of df_prods

In [30]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [31]:
df_prods[df_prods['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [32]:
df_snacks =  df_prods[df_prods['department_id']==19]

In [33]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

### other way to subsetting

In [34]:
df_snacks_2 =  df_prods.loc[df_prods['department_id']==19]

In [35]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [36]:
df_snacks_3 =  df_prods.loc[df_prods['department_id'].isin([19])]

# Question 2: Changing Variable in df_ords

In [43]:
# user_id column in table df_ords is numerical. But this is not needed to be numerical, so we convert into string
df_ords_2['user_id'].astype('str')

0               1
1               1
2               1
3               1
4               1
            ...  
3421078    206209
3421079    206209
3421080    206209
3421081    206209
3421082    206209
Name: user_id, Length: 3421083, dtype: object

# Question 3: Renaming of variable df_ords

In [44]:
## renaming order_hour_of _day to time_of_day in df_ords
df_ords_2.rename(columns = {'order_hour_of_day' : 'time_of_day'}, inplace = True)

In [45]:
#checking the dataframe to see the new column name
df_ords_2.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0  2539329        1             1                   2            8   
1  2398795        1             2                   3            7   
2   473747        1             3                   3           12   
3  2254736        1             4                   4            7   
4   431534        1             5                   4           15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Question 4: Finding the frequency of busiest hour

In [46]:
df_ords_2['time_of_day'].value_counts()

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: time_of_day, dtype: int64

#### The busiest hour for placing order is 10 am.

# Question 5: meaning of value 4 in df_prods

In [51]:
# looking subset of df_prods data with department_id =4
df_prods[df_prods['department_id']==4]

product_id                                      product_name  aisle_id  \
30             31                                White Pearl Onions       123   
42             43                               Organic Clementines       123   
44             45                                 European Cucumber        83   
65             66                         European Style Spring Mix       123   
88             89                    Yogurt Fruit Dip Sliced Apples       123   
...           ...                                               ...       ...   
49582       49578                                Black Garlic Bulbs       123   
49623       49619                                        Opo Squash        83   
49639       49635  Baby Food Blueberry, Parsnip & Buckwheat Stage 2        83   
49661       49657                                 Cabernet Tomatoes        83   
49687       49683                                    Cucumber Kirby        83   

       department_id  prices  
30                 4     7.5  
42                 4    11.5  
44                 4    14.3  
65                 4    11.7  
88                 4    12.6  
...              ...     ...  
49582              4     8.0  
49623              4    12.7  
49639              4    12.5  
49661              4     8.3  
49687              4    13.2  

[1684 rows x 5 columns]

In [52]:
#using the department data dictionary to know the department of department id 4
print(data_dict.get('4'))

{'department': 'produce'}


#### department_id 4 is produce department

# Question 6: Creating a subset

In [59]:
#Based on data dictionary, breakfast has department_id of 14. So creating subset based on departmen_id 14
df_breakfast = df_prods[df_prods['department_id']==14]

In [60]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# Question 7: Creating subset of based on multiple departments

In [64]:
# 5: alcohol, 7= beverages, 12= meat/seafood, 20= deli
df_dinner =  df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [63]:
df_dinner.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

# Question 8: Dataframe of df_dinner

In [65]:
df_dinner.shape

(7650, 5)

#### Dataframe has 7650 rows

# Question 9: Etracting information of user_id of 1

In [69]:
df_user_1 = df_ords_2[df_ords_2['user_id']==1]

In [70]:
df_user_1.head()

order_id  user_id  order_number  orders_day_of_week  time_of_day  \
0  2539329        1             1                   2            8   
1  2398795        1             2                   3            7   
2   473747        1             3                   3           12   
3  2254736        1             4                   4            7   
4   431534        1             5                   4           15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# Question 10: Details of user_id = 1

In [71]:
df_user_1.describe()

user_id  order_number  orders_day_of_week  time_of_day  \
count     11.0     11.000000           11.000000    11.000000   
mean       1.0      6.000000            2.636364    10.090909   
std        0.0      3.316625            1.286291     3.477198   
min        1.0      1.000000            1.000000     7.000000   
25%        1.0      3.500000            1.500000     7.500000   
50%        1.0      6.000000            3.000000     8.000000   
75%        1.0      8.500000            4.000000    13.000000   
max        1.0     11.000000            4.000000    16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

# Question 12: Exporting to prepared data folder

In [75]:
df_ords_2.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [76]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))